pkg/controller/istiocontrolplane/listeners.go

```diff
@@ -0,0 +1,264 @@
+package istiocontrolplane
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"istio.io/operator/pkg/apis/istio/v1alpha1"
+	"istio.io/operator/pkg/helmreconciler"
+)
+
+const (
+	// ChartOwnerKey is the annotation key used to store the name of the chart that created the resource
+	ChartOwnerKey = MetadataNamespace + "/chart-owner"
+)
+
+// IstioListenerFactory is the factory used to create the rendering listeners for IstioControlPlane resources
+type IstioListenerFactory struct{}
+
+var _ helmreconciler.RenderingListenerFactory = &IstioListenerFactory{}
+
+// IstioRenderingListener is a RenderingListener specific to IstioControlPlane resources
+type IstioRenderingListener struct {
+	*helmreconciler.CompositeRenderingListener
+}
+
+// NewRenderingListener returns a new IstioRenderingListener, which is a composite that includes IstioStatusUpdater
+// and IstioChartCustomizerListener.
+func (f *IstioListenerFactory) NewRenderingListener(instance runtime.Object) (helmreconciler.RenderingListener, error) {
+	if istioControlPlane, ok := instance.(*v1alpha1.IstioControlPlane); ok {
```

Comment:
As above, recommend requiring the caller to pass in *v1alpha1.IstioControlPlane

Reply:
maybe this is possible after consolidating the factory interfaces into one.  that said, the helmreconciler bit is intended to be completely generic (hence my comment above about templates/generics).